include/deal.II/fe/fe_enriched.h

```diff
@@ -0,0 +1,585 @@
+ * $ U_i $ and $ U_{jk} $ are the standard and enriched DoFs. This allows to
+ * include in the finite element space a priori knowledge about the partial
+ * differential equation being solved which in turn improves the local
+ * approximation properties of the spaces. This can be useful for highly oscillatory
+ * solutions, problems with domain corners or on unbounded domains or sudden
+ * changes of boundary conditions. PUM method uses finite element spaces which
+ * satisfy the partition of unity property (e.g. FE_Q). Among other properties
+ * this makes the resulting space to reproduce enrichment functions exactly.
+ *
+ *
+ * The simplest constructor of this class takes two finite element objects and an
+ * enrichment function to be used. For example
+ *
+ * @code
+ * FE_Enriched<dim> fe(FE_Q<dim>(2),
+ *                     FE_Q<dim>(1),
+ *                     function)
+ * @endcode
+ *
+ * In this case, standard DoFs are distributed by <code>FE_Q<dim>(2)</code>,
+ * whereas enriched DoFs are coming from a single finite element
+ * <code>FE_Q<dim>(1)</code> used with a single enrichment function
+ * <code>function</code>. In this case, the total number of DoFs on the
+ * enriched element is the sum of DoFs from <code>FE_Q<dim>(2)</code> and
+ * <code>FE_Q<dim>(1)</code>.
+ *
+ * As an example of enrichment function, consider $ \exp(-x) $, which
+ * leads to the following shape functions on unit element:
+ * <table>
+ *   <tr>
+ *     <td align="center">
+ *       @image html fe_enriched_1d.png
+ *     </td>
+ *     <td align="center">
+ *       @image html fe_enriched_h-refinement.png
+ *     </td>
+ *   </tr>
+ *   <tr>
+ *     <td align="center">
+ *       1d element, base and enriched shape functions.
+ *     </td>
+ *     <td align="center">
+ *       enriched shape function corresponding to the central vertex.
+ *     </td>
+ *   </tr>
+ * </table>
+ *
+ * <h3>Using enriched and non-enriched FEs together</h3>
+ *
+ * In most applications it is beneficial to introduce enrichments only in
+ * some part of the domain (e.g. around a crack tip) and use standard FE (e.g. FE_Q)
+ * elsewhere. In order to make the resulting FE space $C^0$ continuous in the
+ * case of continuous enrichmenet functions, all enriched DoFs on the interface
+ * between enriched and non-enriched elements are constrained to be zero.
+ * To get those constraints working automatically, the user needs to wrap a FE
+ * to be used in part of the domain without enrichment:
+ * @code
+ *   FE_Enriched<dim> fe_non_enriched(FE_Q<dim>(1));
+ * @endcode
+ * This constructor is equivalent to calling
+ * @code
+ *   FE_Enriched<dim> fe_non_enriched(FE_Q<dim>(1),
+ *                                    FE_Nothing<dim>(1,true),
+ *                                    NULL);
+ * @endcode
+ * and will result in the correct constraints for enriched DoFs attributed to
+ * support points on the interface between the two regions.
```

Comment:
I think you want to mention the word "hp" here somewhere. Maybe something like this after the first sentence "...elsewhere.":
```
  This can be achieved by using the hp finite element framework in deal.II that allows for the use of different
  elements on different cells. To make the resulting space $C^0$ continuous, it is then necessary for
  the DoFHandler class and DoFTools::make_hanging_node_constraints() function to be able to figure
  out what to do at the interface between enriched and non-enriched cells. Specifically, we want the
  degrees of freedom corresponding to enriched shape functions to be zero at these interfaces. These
  classes and functions can not to do this automatically, but the effect can be achieved by using not
  just a regular FE_Q on cells without enrichment, but to wrap the FE_Q into an FE_Enriched object
  <i>without actually enriching it</i>. This can be done as follows:
```
and then just continue with the code example.